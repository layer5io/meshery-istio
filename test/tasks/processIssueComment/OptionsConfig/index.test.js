const OptionsConfig = require('../../../../src/tasks/processIssueComment/OptionsConfig')

describe('ContentFiles', () => {
    const mockRepository = {
        repo: 'all-contributors-bot-test',
        owner: 'all-contributors',
    }

    test(`Add's new contributor`, async () => {
        const optionsConfig = new OptionsConfig({
            repository: mockRepository,
        })

        optionsConfig.options = {
            contributors: [
                {
                    login: 'jakebolam',
                    name: 'Jake Bolam',
                    avatar_url:
                        'https://avatars2.githubusercontent.com/u/3534236?v=4',
                    profile: 'https://jakebolam.com',
                    contributions: ['code', 'ideas', 'infra', 'test'],
                },
            ],
        }

        await optionsConfig.addContributor({
            login: 'tbenning',
            contributions: ['design'],
            name: 'Tyler Benning',
            avatar_url: 'https://tylerbenning.com/profile.png',
            profile: 'https://tylerbenning.com',
        })

        expect(optionsConfig.options).toMatchSnapshot()
    })

    test(`Add's new contributions for contributor`, async () => {
        const optionsConfig = new OptionsConfig({
            repository: mockRepository,
        })

        optionsConfig.options = {
            contributors: [
                {
                    login: 'jakebolam',
                    name: 'Jake Bolam',
                    avatar_url:
                        'https://avatars2.githubusercontent.com/u/3534236?v=4',
                    profile: 'https://jakebolam.com',
                    contributions: ['code', 'ideas', 'infra', 'test'],
                },
            ],
        }

        await optionsConfig.addContributor({
            login: 'jakebolam',
            contributions: ['doc'],
            name: 'Jake Bolam',
            avatar_url: 'https://avatars2.githubusercontent.com/u/3534236?v=4',
            profile: 'https://jakebolam.com',
        })

        expect(optionsConfig.options).toEqual({
            contributors: [
                {
                    login: 'jakebolam',
                    name: 'Jake Bolam',
                    avatar_url:
                        'https://avatars2.githubusercontent.com/u/3534236?v=4',
                    profile: 'https://jakebolam.com',
                    contributions: ['code', 'ideas', 'infra', 'test', 'doc'],
                },
            ],
        })
    })

    test(`Add's duplicate contributions for contributor`, async () => {
        const optionsConfig = new OptionsConfig({
            repository: mockRepository,
        })

        optionsConfig.options = {
            contributors: [
                {
                    login: 'jakebolam',
                    name: 'Jake Bolam',
                    avatar_url:
                        'https://avatars2.githubusercontent.com/u/3534236?v=4',
                    profile: 'https://jakebolam.com',
                    contributions: ['code', 'ideas', 'infra', 'test'],
                },
            ],
        }

        await optionsConfig.addContributor({
            login: 'jakebolam',
            contributions: ['doc', 'code', 'ideas'],
            name: 'Jake Bolam',
            avatar_url: 'https://avatars2.githubusercontent.com/u/3534236?v=4',
            profile: 'https://jakebolam.com',
        })

        expect(optionsConfig.options).toEqual({
            contributors: [
                {
                    login: 'jakebolam',
                    name: 'Jake Bolam',
                    avatar_url:
                        'https://avatars2.githubusercontent.com/u/3534236?v=4',
                    profile: 'https://jakebolam.com',
                    contributions: ['code', 'ideas', 'infra', 'test', 'doc'],
                },
            ],
        })
    })

    test(`If profile URL is missing protocol, add it for them`, async () => {
        const optionsConfig = new OptionsConfig({
            repository: mockRepository,
        })
        optionsConfig.init()

        await optionsConfig.addContributor({
            login: 'jakebolam',
            contributions: ['ideas'],
            name: 'Jake Bolam',
            avatar_url: 'https://avatars2.githubusercontent.com/u/3534236?v=4',
            profile: 'jakebolam.com',
        })

        expect(optionsConfig.options.contributors).toEqual([
            {
                login: 'jakebolam',
                name: 'Jake Bolam',
                avatar_url:
                    'https://avatars2.githubusercontent.com/u/3534236?v=4',
                profile: 'http://jakebolam.com',
                contributions: ['ideas'],
            },
        ])
    })

    test(`Inits the contributor file`, () => {
        const optionsConfig = new OptionsConfig({
            repository: mockRepository,
        })

        expect(optionsConfig.options).toBeUndefined()

        optionsConfig.init()

        expect(optionsConfig.options).toEqual({
            projectName: mockRepository.repo,
            projectOwner: mockRepository.owner,
            repoType: 'github',
            skipCi: true,
            repoHost: 'https://github.com',
            files: ['README.md'],
            imageSize: 100,
            commit: false,
            contributors: [],
            contributorsPerLine: 7,
        })
    })

    test(`Fetches the contributor file`, async () => {
        const expetectedOptions = {
            projectName: 'my-repo',
            projectOwner: 'my-owner',
            repoType: 'github',
            skipCi: true,
            repoHost: 'https://github.com',
            files: ['README.md'],
            imageSize: 100,
            commit: false,
            contributors: [],
            contributorsPerLine: 7,
        }

        const optionsConfig = new OptionsConfig({
            repository: {
                owner: 'my-owner',
                repo: 'my-repo',
                getFile: async function() {
                    return {
                        content: JSON.stringify(expetectedOptions),
                    }
                },
            },
        })

        expect(optionsConfig.options).toBeUndefined()

        await optionsConfig.fetch()

        expect(optionsConfig.options).toEqual(expetectedOptions)
    })

    test(`Fetches a incorrect contributor file`, async () => {
        const wrongOptions = {
            projectName: 'sdfkjgn',
            projectOwner: 'asdlkjfn',
            repoType: 'sdlkjfn',
            skipCi: true,
            repoHost: 'lkjdghfkjdsfh',
            files: ['README.md'],
            imageSize: 100,
            commit: false,
            contributors: [],
            contributorsPerLine: 7,
        }

        const expetectedOptions = {
            projectName: 'my-repo',
            projectOwner: 'my-owner',
            repoType: 'github',
            skipCi: true,
            repoHost: 'https://github.com',
            files: ['README.md'],
            imageSize: 100,
            commit: false,
            contributors: [],
            contributorsPerLine: 7,
        }

        const optionsConfig = new OptionsConfig({
            repository: {
                owner: 'my-owner',
                repo: 'my-repo',
                getFile: async function() {
                    return {
                        content: JSON.stringify(wrongOptions),
                    }
                },
            },
        })

        expect(optionsConfig.options).toBeUndefined()

        await optionsConfig.fetch()

        expect(optionsConfig.options).toEqual(expetectedOptions)
    })
})
