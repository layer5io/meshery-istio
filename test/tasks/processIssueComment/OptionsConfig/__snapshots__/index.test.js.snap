// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentFiles Add's new contributor 1`] = `
Object {
  "contributors": Array [
    Object {
      "avatar_url": "https://avatars2.githubusercontent.com/u/3534236?v=4",
      "contributions": Array [
        "code",
        "ideas",
        "infra",
        "test",
      ],
      "login": "jakebolam",
      "name": "Jake Bolam",
      "profile": "https://jakebolam.com",
    },
    Object {
      "avatar_url": "https://tylerbenning.com/profile.png",
      "contributions": Array [
        "design",
      ],
      "login": "tbenning",
      "name": "Tyler Benning",
      "profile": "https://tylerbenning.com",
    },
  ],
}
`;
